fedimint-atomic-broadcast/src/lib.rs

```diff
@@ -0,0 +1,246 @@
+//! not save the decisions for items processed before the crash to disk so
+//! Fedimint Consensus either has to keep track of the decisions on disk or
+//! recompute them. I think the simplest solution to this is to use a single
+//! database transaction for the entire session for all consensus critical state
+//! and maintain the index of the current session as part of the state machine.
+//! If we crash mid session all intermediate state changes are lost and we
+//! simply process the returned ordered items again without any recovery
+//! specific logic. At the end of a session all state machines will arrive at
+//! the same state regardless of the exact number of ordered items they
+//! processed to arrive there. Hence, when we receive the first ordered item of
+//! the next session which indicates that the current session is complete we can
+//! bump the index of current session and finally commit the transaction.
+
+mod broadcast;
+mod conversion;
+mod data_provider;
+mod db;
+mod finalization_handler;
+mod keychain;
+mod network;
+mod session;
+mod spawner;
+
+use bitcoin::merkle_tree;
+use bitcoin_hashes::{sha256, Hash};
+/// This function runs the broadcast until a shutdown is intiated either via the
+/// shutdown sender or the [OrderedItem] receiver is dropped.
+pub use broadcast::run;
+use fedimint_core::encoding::{Decodable, Encodable};
+use fedimint_core::PeerId;
+/// This keychain implements naive threshold schnorr signatures over secp256k1.
+/// The broadcasts uses this keychain to sign messages for peers and create
+/// the threshold signatures for the signed blocks.
+pub use keychain::Keychain;
+use tokio::sync::oneshot;
+
+/// The majority of these messages need to be delivered to the intended
+/// [Recipient] in order for the broadcast to make progress. However, the
+/// broadcast does not assume a reliable network layer and implements all
+/// necessary retry logic. Therefore, the caller can discard a message
+/// immediatly if its intended recipient is offline.
+#[derive(Clone, Debug, Encodable, Decodable)]
+pub enum Message {
+    NetworkData(Vec<u8>),
+    BlockRequest(u64),
+    Block(SignedBlock),
+}
+
+/// This enum defines the intented destination of a [Message].
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Recipient {
+    Everyone,
+    Peer(PeerId),
+}
+
+/// This enum specifies wether an [OrderedItem] has been accepted or discarded
+/// by Fedimint Consensus.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Decision {
+    Accept,
+    Discard,
+}
+
+/// If two correct nodes obtain two ordered items from the broadcast they
+/// are guaranteed to be in the same order. However, an ordered items is
+/// only guaranteed to be seen by all correct nodes if a correct node decides to
+/// accept it.
+#[derive(Debug)]
+pub struct OrderedItem {
+    pub item: Vec<u8>,
+    pub block_index: u64,
+    pub peer_id: PeerId,
+    pub decision_sender: oneshot::Sender<Decision>,
```

Comment:
This field does not seem to belong here. All other fields seem like pure data regarding item and its ordering, and this one is a handle to a logical resource (channel sender).